python/dgl/runtime/scheduler.py

```diff
@@ -774,61 +812,84 @@ def _gen_send_reduce(
     mfunc_is_list = utils.is_iterable(mfunc)
     rfunc_is_list = utils.is_iterable(rfunc)
 
-    # Create a tmp frame to hold the feature data.
-    # The frame has the same size and schemes of the
-    # node frame.
-    # TODO(minjie): should replace this with an IR call to make the program stateless.
+    # Create a tmp frame to hold the feature data. The frame has the same size
+    # and schemes of the node frame.
+    # TODO(minjie): should replace this with an IR call to make the program
+    # stateless.
     tmpframe = FrameRef(frame_like(dst_node_frame._frame, len(reduce_nodes)))
     var_out = var.FEAT_DICT(data=tmpframe)
 
+    # 1. If either mfunc or rfunc is builtin, generate adjmat, edge mapping and
+    # message mapping
+    if mfunc_is_list or rfunc_is_list:
+        adj, edge_map, nbits = adj_creator()
+
+    # 2. If rfunc is builtin, generate a mapping from recv nodes to consecutive
+    # output id
+    if rfunc_is_list:
+        out_map = out_map_creator(nbits)
+
+    # 3. First try fused message and reduce function
     if mfunc_is_list and rfunc_is_list:
         # builtin message + builtin reducer
-        # analyze v2v spmv
-        spmv_pairs, mfunc, rfunc = spmv.analyze_v2v_spmv(graph, mfunc, rfunc)
-        adj = adj_creator()
-        spmv.gen_v2v_spmv_schedule(adj, spmv_pairs, var_src_nf, var_ef, var_eid, var_out)
+        spmv.gen_v2v_spmv_schedule(graph=adj,
+                                   mfunc=mfunc,
+                                   rfunc=rfunc,
+                                   src_frame=var_src_nf,
+                                   dst_frame=var_dst_nf,
+                                   edge_frame=var_ef,
+                                   out=var_out,
+                                   out_size=len(reduce_nodes),
+                                   edge_map=edge_map,
+                                   out_map=out_map)
+        return var_out
 
-        if len(mfunc) == 0:
-            # All mfunc and rfunc have been converted to v2v spmv.
-            return var_out
+    var_u, var_v = uv_getter()
 
+    # 4. Unable to fuse, then generate message
     if mfunc_is_list:
-        # Two cases:
-        #  - mfunc is builtin while rfunc is UDF.
-        #  - mfunc and rfunc are both builtin but some combinations
-        #    fall through from the v2v spmv analysis.
-        # In both cases, convert the mfunc to UDF.
-        mfunc = BundledFunction(mfunc)
-
-    # generate UDF send schedule
-    var_u, var_v = uv_getter()
-    var_mf = _gen_send(graph, var_src_nf, var_dst_nf, var_ef, var_u, var_v, var_eid, mfunc)
+        # messages are builtin but reduce is UDF
+        # Create a tmp frame to hold the message.
+        n_message = len(var_eid.data)
+        tmp_msg_frame = FrameRef(frame_like(edge_frame._frame, n_message))
+        var_mf = var.FEAT_DICT(data=tmp_msg_frame)
```

Comment:
Does the TODO in L815 apply here and in L919?

Reply:
Yes. Done